lib/rspec/support.rb

```diff
@@ -101,6 +101,17 @@ def self.with_failure_notifier(callable)
       self.failure_notifier = orig_notifier
     end
 
+    def self.warning_notifier=(callable)
+      thread_local_data[:warning_notifier] = callable
```

Comment:
I'm not sure that we want to scope it to a thread like we did with the failure notifier.  After all, the usage of this in rspec-core is to have a global config that's intended to be set once during config time and then be left for the rest of the process.  It would be confusing for users to configure that and then a warning emitted from another thread and have it be printed rather than raise an error like they have configured.

For the failure notifier we needed it to be scoped to a thread because we mutate it for one example and then mutate it back, and the notifier tracks failures so there's all this mutable state being dealt with.